_episodes/05-snakemake.md

```diff
@@ -0,0 +1,153 @@
+First we need to **identify the steps**.
+
+We copy-pasted the code from the notebook into a `Snakefile` and put the different steps of the workflow in *rules*.
+
+## Explain
+The `Snakefile` is a superset of Python. Everything that is not a Snakemake rule should be valid Python.
+
+### Rules
+A rule looks as follows
+
+~~~snakemake
+rule <rule name>:
+    <rule specs>
+~~~
+{: .source}
+
+Each rule may specify input and output files and some action that describes how to execute the rule.
+
+### Dependency diagram
+Let's go a bit deeper into rule writing by doing a small exercise.
+
+> ## Challenge: create a dependency diagram
+> In a new directory create the following `Snakefile`
+>
+> ~~~python
+> rule all:
+>     input:
+>         "allcaps.txt"
+>
+> rule generate_message:
+>     output:
+>         "message.txt"
+>     shell:
+>         "echo 'Hello, World!' > {output}"
+>
+> rule shout_message:
+>     input:
+>         "message.txt"
+>     output:
+>         "allcaps.txt"
+>     shell:
+>         "tr [a-z] [A-Z] < {input} > {output}"
+> ~~~
+> {: .source}
+>
+> View the dependency diagram: `snakemake --dag | dot | display`, and run the workflow `snakemake -j1`
+> What is happening? Create a new rule that concatenates `message.txt` and `allcaps.txt` (use the `cat` command).
+> Change the `all` rule to require the new output file. When you rerun the workflow, are all the steps repeated?
+>
+> > ## Solution
+> > ~~~python
+> > rule combine:
+> >     input: "message.txt", "allcaps.txt"
+> >     output: "combined.txt"
+> >     shell:
+> >         "cat {input} > {output}"
+> > ~~~
+> >
+> > Only the `combine` rule is being run (in addition to the `all` rule). The dependency diagram should look like this:
+> >
+> > ![Hello world dependency diagram](../fig/hello-graph.svg)
+> > {: .output}
+> {: .solution}
+{: .challenge}
+
+### Wildcards
+An input or output file may contain wildcards. These are names enclosed in braces like `{wildcard}`. A rule with wildcards is the **equivalent of a function** in a normal programming language. We can disect the following rule:
+
+~~~python
+rule plot_result:
+    input:
+        "data/pot0.npy"
+    output:
+        "data/plot_{timestamp}.png"
+    run:
+        t = int(wildcards.timestamp) / 1000
+        pot = np.load(input[0])
+        ... # complicated matplotlib code
+        fig.savefig(output[0])
+~~~
+
+This rule will match any target file in `data/plot_*.png` where the contents of the wildcard is assigned to `wildcard.timestamp`.
+
+## Action
+With this workflow we can do the following:
+
+### Plot the dependency diagram
+~~~bash
+snakemake --dag | dot | display
+~~~
+
+### Run using a single worker
+~~~bash
+time snakemake -j1
+~~~
+
+### Run using multiple workers
+~~~bash
+time snakemake -j4
+~~~
+
+This should run significantly faster than the single worker case.
+
+# Specifics
+## Cosmic structure
+This workflow is located in the `cosmic_structure` folder in the `parallel-python-workshop` repository.
```

Comment:
The `cosmic_structure` folder currently doesn't exist (or at least I cannot find it). Would you please add it and link it in the text?

Reply:
I have this folder visible as "added 21 hours ago" ;)

Reply:
Solved https://github.com/carpentries-incubator/lesson-parallel-python/pull/45/commits/3e2dbdeacc6f15d87eb3ebd118057113a6cd723a